### ISSUE TITLE
Functional Programming In Java

### ISSUE THREAD

**Issue by john-amiscaray** [section-io/engineering-education#459]
### Brief Summary: 
This article will be a guide on functional programming in Java. This includes how it works, what support Java provides for it, and practical use cases for it.

### Key Takeaways: 
- What is functional programming?
- What support does Java provide for it?
- Lambda Expressions
- Stream api
- Method reference

### References: 
https://medium.com/better-programming/functional-programming-with-java-an-introduction-daa783355731
https://dzone.com/articles/functional-programming-patterns-with-java-8
https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html

**Comment by ninjaginja** [comment]
@john-amiscaray Thanks for the topic suggestion. Are you able to expand upon how your article will differ from the references that you've listed?

**Comment by john-amiscaray** [comment]
@ninjaginja The article I plan to write will be more beginner friendly approach as it will take the perspective of teaching someone who has ever learned about these concepts before. This will be done using some analogy and comparison of these concepts to other concepts in Java or in the real world to give the reader a deeper understanding of these concepts. The other articles on the other hand have parts that are much faster pace, may not be as easy to understand as a beginner, or may assume you understand some of the concepts. As well, unlike the other articles, I plan to go into detail about how lambda expressions work in Java under the hood. I have mostly used these sources just to make sure what I am saying is correct as possible, and in one cases to just refresh myself on the concepts, or for inspiration for some of the examples I plan to provide.

**Comment by ninjaginja** [comment]
This is great @john-amiscaray . Thanks for the extended detail!

**Comment by john-amiscaray** [comment]
@ninjaginja I just realized one of the articles I referenced does have a very brief explanation on how lambda expressions work under the hood. That is a mistake in my response that I just realized as I glossed over that part of my source. However their explanation is very brief and not as in depth as the explanation on it that I provided, and the explanation I made was not based on the explanation they provided.
